app/services/record_declaration.rb

```diff
@@ -0,0 +1,170 @@
+    self.raw_declaration_date = raw_declaration_date
+    super
+  end
+
+  def participant_identity
+    @participant_identity ||= ParticipantIdentity.find_by(external_identifier: participant_id)
+  end
+
+  def participant_profile
+    @participant_profile ||= ParticipantProfileResolver
+                               .call(
+                                 participant_identity:,
+                                 course_identifier:,
+                                 cpd_lead_provider:,
+                               )
+  end
+
+  def milestone
+    return unless schedule
+
+    schedule.milestones.find_by(declaration_type:)
+  end
+
+  def schedule
+    participant_profile&.schedule
+  end
+
+private
+
+  attr_writer :raw_declaration_date
+
+  def participant_profile_for_course_identifier
+    return unless participant_identity
+  end
+
+  def declaration_attempt
+    if participant_id
+      @declaration_attempt ||= ParticipantDeclarationAttempt.create!(
+        course_identifier:,
+        declaration_date:,
+        declaration_type:,
+        cpd_lead_provider:,
+        user_id: participant_id,
+      )
+    end
+  end
+
+  def set_eligibility
+    if participant_declaration.duplicate_declarations.any?
+      participant_declaration.update!(superseded_by: original_participant_declaration)
+      participant_declaration.make_ineligible!(reason: :duplicate)
+    elsif participant_profile.fundable?
+      participant_declaration.make_eligible!
+    end
+  end
+
+  def validate_milestone_exists
+    if milestone.blank?
+      errors.add(:declaration_type, I18n.t(:mismatch_declaration_type_for_schedule))
+    end
+  end
+
+  def participant_declaration
+    @participant_declaration ||= participant_profile
+                                   .participant_declarations
+                                   .submitted
+                                   .or(
+                                     participant_profile
+                                       .participant_declarations
+                                       .billable,
+                                   )
+                                   .find_or_create_by!(declaration_parameters)
+                                   .tap { |pd| pd.update!(uplift_flags) }
+  end
+
+  def uplift_flags
+    {
+      pupil_premium_uplift: participant_profile.pupil_premium_uplift,
+      sparsity_uplift: participant_profile.sparsity_uplift,
+    }
+  end
+
+  def declaration_parameters
+    {
+      course_identifier:,
+      declaration_date:,
+      declaration_type:,
+      cpd_lead_provider:,
+      user: participant_identity.user,
+    }
+  end
+
+  def validate_evidence_held?
+    declaration_type.present? && declaration_type != "started"
+  end
+
+  def original_participant_declaration
+    @original_participant_declaration ||= participant_declaration.duplicate_declarations.order(created_at: :asc).first
+  end
+
+  def validates_billable_slot_available
```

Comment:
Is this method called? I can't see from where.

Reply:
the implementation has also changed. i'm not sure the new implementation does what it was intended to do? `participant_declaration` looks to return an `ActiveRecord` object, the code then chains `where` which I think is not valid

Reply:
re-ad